File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        zeros = 0
        j = 0
        i = 0
        res = 0
        if sum(nums) == 0 and k==0:
            return 0
        while j < len(nums):
            if nums[j] == 0:
                zeros += 1
            while i < j and zeros > k:
                if nums[i] == 0:
                    zeros -= 1
                i += 1
            res = max(res, j-i+1)
            j += 1
        return res
                

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int i=0;
        int count = 0;
        int res = 0;
        for(int j=0;j<nums.length;j++){
            count += nums[j];
            while(i < j && j-i+1 > count + k){
                count -= nums[i];
                i++;
            }
            if(k != 0 || count > 0)
                res = Math.max(res, j-i+1);  
        }
        return res;
    }
}